player/javascript/defaults.js

```diff
@@ -0,0 +1,495 @@
+        for (var len = handlers.length, i = 0; i < len; i++) {
+            var cb = handlers[i].cb;  // 'handlers' won't mutate, but unregister
+            if (cb)                   // could remove cb from some items
+                cb(e);
+        }
+    }
+}
+
+//  ----- property observers -----
+var next_oid = 1,
+    observers = new_cache();  // items of id: fn
+
+mp.observe_property = function(name, format, fn) {
+    var id = next_oid++;
+    observers[id] = fn;
+    return mp._observe_property(id, name, format || undefined);  // allow null
+}
+
+mp.unobserve_property = function(fn) {
+    for (var id in observers) {
+        if (observers[id] == fn) {
+            delete observers[id];
+            mp._unobserve_property(id);
+        }
+    }
+}
+
+function notify_observer(e) {
+    var cb = observers[e.id];
+    if (cb)
+        cb(e.name, e.data);
+}
+
+//  -----  Client messages -----
+var messages = new_cache();  // items of name: fn
+
+// overrides name. no libmpv API to reg/unreg specific messages.
+mp.register_script_message = function(name, fn) {
+    messages[name] = fn;
+}
+
+mp.unregister_script_message = function(name) {
+    delete messages[name];
+}
+
+function dispatch_message(ev) {
+    var cb = ev.args.length ? messages[ev.args[0]] : false;
+    if (cb)
+        cb.apply(null, ev.args.slice(1));
+}
+
+//  ----- hooks -----
+var next_hid = 1,
+    hooks = new_cache();  // items of id: fn
+
+function hook_run(id, cont) {
+    var cb = hooks[id];
+    if (cb)
+        cb();
+    mp.commandv("hook_ack", cont);
+}
+
+mp.add_hook = function add_hook(name, pri, fn) {
+    if (next_hid == 1)  // doesn't really matter if we do it once or always
+        mp.register_script_message("hook_run", hook_run);
+    var id = next_hid++;
+    hooks[id] = fn;
+    return mp.commandv("hook_add", name, id, pri);
+}
+
+/**********************************************************************
+ *  key bindings
+ *********************************************************************/
+// binds: items of (binding) name which are objects of:
+// {cb: fn, forced: bool, maybe input: str, repeatable: bool, complex: bool}
+var binds = new_cache();
+
+function dispatch_key_binding(name, state) {
+    var cb = binds[name] ? binds[name].cb : false;
+    if (cb)  // "script-binding [<script_name>/]<name>" command was invoked
+        cb(state);
+}
+
+function update_input_sections() {
+    var def = [], forced = [];
+    for (var n in binds)  // Array.join() will later skip undefined .input
+        (binds[n].forced ? forced : def).push(binds[n].input);
+
+    var sect = "input_" + mp.script_name;
+    mp.commandv("define-section", sect, def.join("\n"), "default");
+    mp.commandv("enable-section", sect, "allow-hide-cursor+allow-vo-dragging");
+
+    sect = "input_forced_" + mp.script_name;
+    mp.commandv("define-section", sect, forced.join("\n"), "force");
+    mp.commandv("enable-section", sect, "allow-hide-cursor+allow-vo-dragging");
+}
```

Comment:
It implies it could need some heavy maintenance. If the mechanism changes, I'd just rewrite the code in the Lua wrapper. Can I expect you'd do the same in the JS one?